### PR TITLE
Fix new face with mesh editing

### PR DESCRIPTION
**Pull request by vcloarec**
This PR is related to #50033, #50035 and  #50037.

Indeed, before this PR, adding new faces with mesh editing suffered of several issues:
- when entered in adding new face mode, to create a new vertex outside the current mesh, user had to double click. A way of digitizing quite different from digitizing vector layer.
- to undo edges of the new face, during its creation, user can press backspace, as when digitizing vector layer. But, here, the new vertices created during this operation was still present as free vertex even after backspace or undo the newly created face. This behavior is quite ugly for the user...
- There was a bug when user, after starting to add a new face, go to another tool and come back to the mesh editing tool: rubber band is broken, and also the general behavior of adding face.

This PR fixes all this issues by changing the approach:
- when entering in "add new face" mode, adding new vertices is simply done by a simple click outside the mesh. The vertices is actually created only after the face is validated by a right click. Backspace undo new edge AND new vertices.
- when the user is not in the "add new face" mode, the user still need a double click to add a vertex outside the mesh (it is a free vertex, not related to any face) and inside the current mesh (a simple vertex related to faces). It is necessary because simple click could interact with other action possible (select element, move, ...) and because adding a new vertex is a strong action that has to be really wanted by the user.

ping @DelazJ 

**BEFORE**
new_face_before

**AFTER**
new_face_after

### CODE REVIEW

File: src/core/mesh/qgsmesheditor.cpp [hunk 1]
```diff
@@ -189,6 +229,60 @@ bool QgsMeshEditor::faceCanBeAdded( const QgsMeshFace &face )
...
+{
+  QgsMeshEditingError error;
+  QList<int> face = prepareFaceWithNewVertices( verticesIndex, newVertices, error );
+
+  if ( error.errorType != Qgis::MeshEditingErrorType::NoError )
+    return false;
+
+  // if we are here, the face is convex and not flat
+
+  // Now we chekc the topology of the potential new face
```
Comment (DelazJ) [inline src/core/mesh/qgsmesheditor.cpp:242]:
```suggestion
  // Now we check the topology of the potential new face
```

File: src/core/mesh/qgsmesheditor.h [hunk 2]
```diff
@@ -88,21 +88,45 @@ class CORE_EXPORT QgsMeshEditor : public QObject
...
-    bool faceCanBeAdded( const QgsMeshFace &face );
+    /**
+     * Returns TRUE if a \a face can be added to the mesh
+     *
+     * \note All vertices related to this face must be already in the mesh.
+     */
+    bool faceCanBeAdded( const QgsMeshFace &face ) const;
+
+    /**
+     * Returns TRUE if a face formed by some vertices con be added to the mesh.
```
Comment (DelazJ) [inline src/core/mesh/qgsmesheditor.h:99]:
```suggestion
     * Returns TRUE if a face formed by some vertices can be added to the mesh.
```

File: src/core/mesh/qgstopologicalmesh.cpp [hunk 3]
```diff
@@ -675,6 +667,33 @@ QgsMeshEditingError QgsTopologicalMesh::counterClockwiseFaces( QgsMeshFace &face
       direction = crossProd / std::fabs( crossProd );
   }
 
+  return QgsMeshEditingError( Qgis::MeshEditingErrorType::NoError, -1 );
+}
+
+QgsMeshEditingError QgsTopologicalMesh::counterClockwiseFaces( QgsMeshFace &face, QgsMesh *mesh )
+{
+  // First check if the topology of the face , then put it counter clockwise if needed
```
Comment (DelazJ) [inline src/core/mesh/qgstopologicalmesh.cpp:675]:
```suggestion
  // First check the topology of the face, then put it counter clockwise if needed
```
Or missing words?

File: src/core/mesh/qgstopologicalmesh.cpp [hunk 4]
```diff
@@ -675,6 +667,33 @@ QgsMeshEditingError QgsTopologicalMesh::counterClockwiseFaces( QgsMeshFace &face
...
+  // If the index are not well ordered (edges intersect), invalid face
```
Comment (DelazJ) [inline src/core/mesh/qgstopologicalmesh.cpp:676]:
```suggestion
  // If the indices are not well ordered (edges intersect), invalid face
```

File: src/core/mesh/qgstopologicalmesh.h [hunk 5]
```diff
@@ -306,6 +306,14 @@ class CORE_EXPORT QgsTopologicalMesh
     //! Checks the topology of the face and sets it counter clockwise if necessary
     static QgsMeshEditingError counterClockwiseFaces( QgsMeshFace &face, QgsMesh *mesh );
 
+    /**
+     * Checks the topology of the \a vertices as the are contained in a face and returns indication on direction.
```
Comment (DelazJ) [inline src/core/mesh/qgstopologicalmesh.h:310]:
```suggestion
     * Checks the topology of the \a vertices as they are contained in a face and returns indication on direction.
```
Or missing words?

**Review by DelazJ — COMMENTED** [review]
Sorry @vcloarec I can't test the branch but the description is as we discussed.
Found just a few typos.

**Review by uclaros — COMMENTED** [review]
This is much more intuitive! :+1:

**Review by uclaros — APPROVED** [review]

**Review by uclaros — APPROVED** [review]

**Comment by DelazJ** [comment]
@vcloarec Tested, it works as expected. Great.
As a buf fix should this not be backported to 3.28?

**Comment by vcloarec** [comment]
@DelazJ  PR for backport done

**Pull request merged**